torch/nested/_internal/nested_tensor.py

```diff
@@ -132,38 +129,75 @@ def __new__(
 
         return r
 
-    def __init__(self, values, offsets, *, lengths=None, **kwargs):
+    def __init__(self, values, metadata, non_contig_offsets=None, **kwargs):
         super().__init__()
 
         self._values = values
-        self._offsets = offsets
-        self._lengths = lengths
-
-        # holds properties that are computed lazily
-        self._metadata_cache = kwargs.get("_metadata_cache") or {}
+        self._non_contig_offsets = non_contig_offsets
 
         # collapsed ragged dim must always be dynamic
         torch._dynamo.maybe_mark_dynamic(self, self._ragged_idx)
         torch._dynamo.maybe_mark_dynamic(self._values, self._ragged_idx - 1)
 
-        # min / max sequence length should be dynamic if present
-        max_seqlen_tensor = self._metadata_cache.get("max_seqlen", None)
-        if max_seqlen_tensor is not None:
-            torch._dynamo.mark_dynamic(max_seqlen_tensor, 0)
-        min_seqlen_tensor = self._metadata_cache.get("min_seqlen", None)
-        if min_seqlen_tensor is not None:
-            torch._dynamo.mark_dynamic(min_seqlen_tensor, 0)
-
-    def values(self):
+    def values(self) -> torch.Tensor:
         # dispatch to get proper view relationship
         return torch._nested_get_values(self)  # type: ignore[attr-defined]
 
-    def offsets(self):
-        return self._offsets
+    # Public APIs for lengths/offsets returns plain tensor for BC reasons, e.g. user
+    # can access .data_ptr() on it, etc.
+    # .lengths() and .offsets() do not automatically do conversion between lengths
+    # and offsets. If the requested variant does not exist, then None is returned.
+    def offsets(self) -> torch.Tensor:
+        ret = self._offsets
+        assert ret is not None
+        return ret
 
-    def lengths(self):
+    def lengths(self) -> Optional[torch.Tensor]:
         return self._lengths
 
+    def __getattr__(self, name) -> Optional[torch.Tensor]:
```

Comment:
I think technically this typing is wrong given get getattr is used for everything related to the class attributes (e.g. `_ragged_idx`)

Reply:
I don't think _ragged_idx goes through this path. We only activate this path after fields have already been checked.

Reply:
That's right, I forgot the details on the distinction between `__getattr__` and `__getattribute__`